posts/improving-the-access-control-of-a-jhipster-application.md

```diff
@@ -0,0 +1,222 @@
+
+The code below from the class `UserResource.java` shows how JHipster does that:
+
+```java
+@GetMapping("/users/{login:" + Constants.LOGIN_REGEX + "}")
+@Timed
+public ResponseEntity<UserDTO> getUser(@PathVariable String login) {
+    log.debug("REST request to get User : {}", login);
+    return ResponseUtil.wrapOrNotFound(
+        userService.getUserWithAuthoritiesByLogin(login)
+            .map(UserDTO::new));
+}
+
+@DeleteMapping("/users/{login:" + Constants.LOGIN_REGEX + "}")
+@Timed
+@PreAuthorize("hasRole(\"" + AuthoritiesConstants.ADMIN + "\")")
+public ResponseEntity<Void> deleteUser(@PathVariable String login) {
+    log.debug("REST request to delete User: {}", login);
+    userService.deleteUser(login);
+    return ResponseEntity.ok().headers(HeaderUtil.createAlert( "userManagement.deleted", login)).build();
+}
+```
+
+Both methods use the url `/api/users/{login}` but with a different HTTP verb (`GET` and `DELETE`). The annotation `@PreAuthorize` makes sure that the user can invoke the method. In the example above, the call to delete an user will fail if the authenticated user is not an **admin**.
+
+The annotation `@PreAuthorize` can be used to have a very granular access-control and solve the problem issued in the introduction.
+
+## Entities owned by an user
+
+One last concept about the backend that might be useful to implement is to have an access-control at the user level. That way a regular user will only have access to its entities and will not be allow to view other users entities. That can be  done at the UI level but that will not protect a malicious user to brute force the API with a range of IDs for example.
+
+In order to implement this logic, a relationship must be created between the entities and the `User` entity. Then using the class `SecurityUtils`, the login of the connected user can be retrieved.
+
+Here is an example of this business implementation using the entity `BankAccount` that has a `ManyToOne` relationship with `User`:
+
+```java
+@GetMapping("/bank-accounts/{id}")
+@Timed
+public ResponseEntity<BankAccountDTO> getBankAccount(@PathVariable Long id) {
+    log.debug("REST request to get BankAccount : {}", id);
+    Optional<BankAccountDTO> bankAccountDTO = bankAccountRepository.findById(id)
+        .map(bankAccountMapper::toDto);
+
+    // Return 404 if the entity is not owned by the connected user
+    Optional<String> userLogin = SecurityUtils.getCurrentUserLogin();
+    if (bankAccountDTO.isPresent() &&
+        userLogin.isPresent() &&
+        userLogin.get().equals(bankAccountDTO.get().getUserLogin())) {
+        return ResponseUtil.wrapOrNotFound(bankAccountDTO);
+    } else {
+        return ResponseEntity.notFound().build();
+    }
+}
+```
+
+With the above code, the user will be only able to retrieve its bank account and will get a 404 response in case of an other user's bank account.
```

Comment:
With the above code, the user will only be able to retrieve their own bank account and will get a 404 response if they try another user's bank account.